sys/net/gnrc/link_layer/netdev2/gnrc_netdev2_ieee802154.c

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright (C) 2016 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ * @author  Martine Lenders <mlenders@inf.fu-berlin.de>
+ */
+
+#include <stddef.h>
+
+#include "od.h"
+#include "net/gnrc.h"
+#include "net/ieee802154.h"
+
+#include "net/gnrc/netdev2/ieee802154.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#define _TAIL_LEN       (4)
+#define _LQI_TAIL_POS   (2)
+#define _RSSI_TAIL_POS  (3)
+
+static gnrc_pktsnip_t *_recv(gnrc_netdev2_t *gnrc_netdev2);
+static int _send(gnrc_netdev2_t *gnrc_netdev2, gnrc_pktsnip_t *pkt);
+
+int gnrc_netdev2_ieee802154_init(gnrc_netdev2_t *gnrc_netdev2,
+                                 netdev2_ieee802154_t *dev)
+{
+    gnrc_netdev2->send = _send;
+    gnrc_netdev2->recv = _recv;
+    gnrc_netdev2->dev = (netdev2_t *)dev;
+
+    return 0;
+}
+
+static gnrc_pktsnip_t *_make_netif_hdr(uint8_t *mhr)
+{
+    gnrc_pktsnip_t *snip;
+    uint8_t src[IEEE802154_LONG_ADDRESS_LEN], dst[IEEE802154_LONG_ADDRESS_LEN];
+    int src_len, dst_len;
+    le_uint16_t src_pan, dst_pan;   /* TODO: hand-up PAN IDs? */
+
+    dst_len = ieee802154_get_dst(mhr, dst, &dst_pan);
+    src_len = ieee802154_get_src(mhr, src, &src_pan);
+    if ((dst_len < 0) || (src_len < 0)) {
+        DEBUG("_make_netif_hdr: unable to get addresses\n");
+        return NULL;
+    }
+    /* allocate space for header */
+    snip = gnrc_netif_hdr_build(src, (size_t)src_len, dst, (size_t)dst_len);
+    if (snip == NULL) {
+        DEBUG("_make_netif_hdr: no space left in packet buffer\n");
+        return NULL;
+    }
+    /* set broadcast flag for broadcast destination */
+    if ((dst_len == 2) && (dst[0] == 0xff) && (dst[1] == 0xff)) {
+        gnrc_netif_hdr_t *hdr = snip->data;
+        hdr->flags |= GNRC_NETIF_HDR_FLAGS_BROADCAST;
+    }
+    return snip;
+}
+
+static gnrc_pktsnip_t *_recv(gnrc_netdev2_t *gnrc_netdev2)
+{
+    netdev2_ieee802154_t *dev = (netdev2_ieee802154_t *)gnrc_netdev2->dev;
+    gnrc_pktsnip_t *pkt = NULL;
+    int bytes_expected = dev->driver->recv((netdev2_t *)dev, NULL, 0);
+
+    if (bytes_expected > 0) {
+        int nread;
+
+        pkt = gnrc_pktbuf_add(NULL, NULL, bytes_expected, GNRC_NETTYPE_UNDEF);
+        if (pkt == NULL) {
+            DEBUG("_recv_ieee802154: cannot allocate pktsnip.\n");
+            return NULL;
```

Comment:
As this hopefully won't happen too often, a set() option is probably the way to go, something like "NETDEV2_DISCARD_PACKET".